src/providers/db2/qgsdb2featureiterator.cpp

```diff
@@ -0,0 +1,332 @@
+
+    mStatement += QString( ",%1" ).arg( fieldname );
+
+    mAttributesToFetch.append( i );
+  }
+
+  // get geometry col in WKB format
+  if ( !( request.flags() & QgsFeatureRequest::NoGeometry ) && mSource->isSpatial() )
+  {
+    mStatement += QString( ",VARCHAR(DB2GSE.ST_ASTEXT(%1)) AS %1 " ).arg( mSource->mGeometryColName );
+//    mStatement += QString( ",DB2GSE.ST_ASTEXT(%1) AS %1 " ).arg( mSource->mGeometryColName );
+
+    mAttributesToFetch.append( 2 );
+  }
+
+  mStatement += QString( " FROM %1.%2" ).arg( mSource->mSchemaName, mSource->mTableName );
+
+  bool filterAdded = false;
+  // set spatial filter
+  if ( !request.filterRect().isNull() && mSource->isSpatial() && !request.filterRect().isEmpty() )
+  {
+    QString r;
+    QTextStream foo( &r );
+
+    /**
+     * The first form (def USE_POLY) mirrors Microsoft SQL's implementation. //TODO test this form
+     * The second form (not def USE_POLY) is more efficient for DB2 because it
+     * compares envelopes of the geometries with the envelope of the viewing
+     * rectangle instead of an edge-by-edge geometric comparison of the geometry
+     * and the viewing rectangle. This has the drawback that it can sometimes
+     * include geometries that are completely outside the viewing rectangle but
+     * this is generally not an issue.
+     */
+#ifdef USE_POLY
+    foo.setRealNumberPrecision( 8 );
+    foo.setRealNumberNotation( QTextStream::FixedNotation );
+    foo <<  qgsDoubleToString( request.filterRect().xMinimum() ) << ' ' <<  qgsDoubleToString( request.filterRect().yMinimum() ) << ", "
+    <<  qgsDoubleToString( request.filterRect().xMaximum() ) << ' ' << qgsDoubleToString( request.filterRect().yMinimum() ) << ", "
+    <<  qgsDoubleToString( request.filterRect().xMaximum() ) << ' ' <<  qgsDoubleToString( request.filterRect().yMaximum() ) << ", "
+    <<  qgsDoubleToString( request.filterRect().xMinimum() ) << ' ' <<  qgsDoubleToString( request.filterRect().yMaximum() ) << ", "
+    <<  qgsDoubleToString( request.filterRect().xMinimum() ) << ' ' <<  qgsDoubleToString( request.filterRect().yMinimum() );
+
+    mStatement += QString( " WHERE DB2GSE.ST_INTERSECTS(%1,DB2GSE.ST_POLYGON('POLYGON((%2))',%3)) = 1" ).arg(
+                    mSource->mGeometryColName, r, QString::number( mSource->mSRId ) );
+#else
+    mStatement += QString( " WHERE DB2GSE.ENVELOPESINTERSECT(%1, %2, %3, %4, %5, %6) = 1" ).arg(
+                    mSource->mGeometryColName,
+                    qgsDoubleToString( request.filterRect().xMinimum() ),
+                    qgsDoubleToString( request.filterRect().yMinimum() ),
+                    qgsDoubleToString( request.filterRect().xMaximum() ),
+                    qgsDoubleToString( request.filterRect().yMaximum() ),
+                    QString::number( mSource->mSRId ) );
+
+#endif
+    filterAdded = true;
+  }
+
+  // set fid filter
+  if ( request.filterType() == QgsFeatureRequest::FilterFid && !mSource->mFidColName.isEmpty() )
```

Comment:
I'd *strongly* suggest adding handling of FilterFids requests too.. this is a huge performance win across lots of different operations